bottlecap/src/telemetry/listener.rs

```diff
@@ -30,22 +47,26 @@ impl HttpRequestParser {
     /// It will also error if the headers cannot be parsed.
     ///
     /// Or if the body cannot be parsed.
-    pub fn from_stream(mut stream: &TcpStream) -> Result<HttpRequestParser, Box<dyn Error>> {
-        stream.set_nonblocking(true)?;
+    pub async fn from_stream(stream: &mut TcpStream) -> Result<HttpRequestParser, TcpError> {
         let mut parser = HttpRequestParser {
             headers: HashMap::new(),
             body: String::new(),
         };
 
         let mut headers_buf = [0u8; HEADERS_BUFFER_SIZE];
         loop {
-            match stream.read(&mut headers_buf) {
+            // select here
+            match stream.read(&mut headers_buf).await {
+                Ok(0) => {
+                    return Err(TcpError::Close("Connection closed by client".to_string()));
+                }
```

Comment:
```suggestion
            match stream.read(&mut headers_buf).await {
                Ok(0) => Err(TcpError::Close("Connection closed by client".to_string())),
```

Reply:
we want to break out of the loop so we need the return